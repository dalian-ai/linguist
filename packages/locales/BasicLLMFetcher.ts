import OpenAI, { ClientOptions } from 'openai';

import { LLMFetcher, MessageObject } from './LLMFetcher';

export class BasicLLMFetcher implements LLMFetcher {
	private readonly api: OpenAI;
	constructor(
		params: ClientOptions,
		private readonly config: {
			model: string;
			temperature?: number;
			systemPrompt?: MessageObject[];
		},
	) {
		this.api = new OpenAI(params);
	}

	async query(messages: MessageObject[]): Promise<MessageObject[]> {
		const response = await this.api.chat.completions.create({
			model: this.config.model,
			temperature: this.config.temperature,
			messages: [...(this.config.systemPrompt ?? []), ...messages],
		});

		const result: MessageObject[] = [];
		for (const choice of response.choices) {
			const { message } = choice;
			if (typeof message.content !== 'string') throw new Error('Invalid response');

			result.push(message as MessageObject);
		}

		return result;
	}

	getLengthLimit() {
		return 25000;
	}

	getRequestsTimeout() {
		return 300;
	}
}
