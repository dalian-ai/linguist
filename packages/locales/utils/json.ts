export const sliceJsonString = (sourceJson: string, maxLength: number) => {
	const json = JSON.parse(sourceJson);
	if (typeof json !== 'object' || json === null)
		throw new TypeError('Json data cannot be split to slices');

	const result: Array<Array<[string, unknown]>> = [];
	let offset = 0;
	for (const slice of Object.entries(json)) {
		const isEmptyChunk = result[offset] && result[offset].length > 0;

		// Fill immediately for empty chunks
		if (!isEmptyChunk) {
			result[offset] = [slice];
			continue;
		}

		// Fill if fit
		const candidate = [...result[offset], slice];
		if (JSON.stringify(Object.fromEntries(candidate)).length <= maxLength) {
			result[offset] = candidate;
			continue;
		}

		// Fill next chunk
		result[++offset] = [slice];
		continue;
	}

	return result.map((entries) => JSON.stringify(Object.fromEntries(entries)));
};

export function getType(value: unknown): string {
	if (Array.isArray(value)) return 'array';
	if (value === null) return 'null';
	return typeof value;
}

export function isEqualStructures(a: unknown, b: unknown): boolean {
	const typeA = getType(a);
	const typeB = getType(b);
	if (typeA !== typeB) return false;

	if (typeA === 'object') {
		const keysA = Object.keys(a as object);
		const keysB = Object.keys(b as object);
		if (keysA.length !== keysB.length) return false;
		keysA.sort();
		keysB.sort();
		for (let i = 0; i < keysA.length; i++) {
			if (keysA[i] !== keysB[i]) return false;
		}
		for (const key of keysA) {
			if (
				!isEqualStructures(
					(a as Record<string, unknown>)[key],
					(b as Record<string, unknown>)[key],
				)
			) {
				return false;
			}
		}
		return true;
	}

	if (typeA === 'array') {
		const arrA = a as unknown[];
		const arrB = b as unknown[];
		if (arrA.length !== arrB.length) return false;
		for (let i = 0; i < arrA.length; i++) {
			if (!isEqualStructures(arrA[i], arrB[i])) return false;
		}
		return true;
	}

	// For primitive types: types must match, values don't matter
	return true;
}
