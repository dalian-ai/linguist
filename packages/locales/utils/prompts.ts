export const codeBlock = (code: string, language?: string) =>
	['```' + (language ?? ''), code, '```'].join('\n');

export const getJsonTranslationPrompt = (json: string, from: string, to: string) => {
	// use full language name
	const langFormatter = new Intl.DisplayNames(['en'], { type: 'language' });
	const originLang = from == 'auto' ? 'auto' : langFormatter.of(from);
	const targetLang = langFormatter.of(to);

	const prettifiedJson = JSON.stringify(JSON.parse(json), null, 2);

	return `You are a translation service for translate localization files.

	I will provide a JSON string with text, and your purpose is to translate all string values (not keys) from language ${originLang} to language ${targetLang}.

	If I specify the source language as 'auto', you should automatically detect it and translate it into the target language I set.

	The JSON object in your response must have the same structure and length as the one in the request. Do not add any explanations â€” translate strictly according to the content and its context.

	Your response must contain only valid raw JSON text with no any formatting and with no code block.

	# Translation specification

	You must never change any key values in object.
	You can use object keys to understand context.

	You must translate only values in "message" property.
	
	Never translate anything except "message" property, it's just a context to help you understand how this "message" will be used.

	You must never change slogans and marketing descriptions for a products.

	Never change text intention. For example, if text is formulated as an question, you must never remove question mark.

	# The context

	You must consider next project description as context when you work on localization:

	> Linguist is a browser extension for translate web pages. It is a full-featured translation solution.
	> With Linguist user may run full page translation, translate any custom text, translate selected text.
	> Linguist supports a custom translators, so user can add its own implementation of translator module.
	> Linguist is a privacy focused, includes embedded modules for offline translation in user device (locally).

	# Translation recommendations

	You may rephrase texts to make it look native for target language, but you must keep the common sense of every "message" according to its context of use.

	Keep language professional and clear.
	Translate text as an educated person, don't use colloquialisms and explicit language.
	Make sure word forms looks clear and not ridiculous.

	Be careful when creating an JSON object; it must be syntactically correct and do not change quotation marks.

	# Your task

	Here is the JSON to translate:
	${codeBlock(prettifiedJson, 'json')}
	`;
};
