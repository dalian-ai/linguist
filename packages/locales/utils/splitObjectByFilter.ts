import isEqual from 'deep-equal';
import traverse from 'traverse';

export type ObjectFilter<T extends {} = {}> = (context: {
	content: T;
	path: string[];
}) => boolean;

/**
 * Split object by filter to `included` and `excluded` subsets
 */

export const splitObjectByFilter = <T extends {} = {}>(
	object: T,
	filter: ObjectFilter<T>,
) => {
	const includedPaths: string[][] = traverse(object)
		.paths()
		.slice(1)
		.filter((path) =>
			filter({
				content: structuredClone(object),
				path: path,
			}),
		);

	const included = traverse(object).map(function fn() {
		if (this.isRoot) return;

		// Delete not included nodes
		const isIncluded = includedPaths.some((path) => isEqual(path, this.path));
		if (!isIncluded) {
			this.remove();

			// Don't walk nested elements
			this.block();
		}
	});

	const excluded = traverse(object).map(function fn() {
		if (this.isRoot) return;

		// Collect filtered out nodes
		const isIncluded = includedPaths.some((path) => isEqual(path, this.path));
		if (isIncluded) {
			this.remove();
		} else {
			// Left whole node with all content
			this.block();
		}
	});

	return { included, excluded };
};
