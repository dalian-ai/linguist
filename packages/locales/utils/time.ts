export function getWaitTimeWithJitter({
	base,
	max,
	retry,
}: {
	base: number;
	max: number;
	retry: number;
}): number {
	const exp = Math.min(base * 2 ** retry, max);
	const jitter = Math.random() * exp * 0.5; // 0â€“50% jitter
	return exp + jitter;
}

export const waitTimeWithJitter = (
	...options: Parameters<typeof getWaitTimeWithJitter>
) => new Promise((res) => setTimeout(res, getWaitTimeWithJitter(...options)));
