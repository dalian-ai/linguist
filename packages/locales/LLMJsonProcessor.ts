import deepmerge from 'deepmerge';

import { LLMFetcher, MessageObject } from './LLMFetcher';
import { sliceJsonString } from './utils/json';
import { ObjectFilter, splitObjectByFilter } from './utils/splitObjectByFilter';
import { waitTimeWithJitter } from './utils/time';

export type ParsingErrorFixer = (response: string) => MessageObject[];

export type ValidatorResult =
	| {
			isValid: true;
	  }
	| {
			isValid: false;
			reason?: unknown;
			correctionRequest?: MessageObject[];
	  };

export type ObjectTransformingValidator = (
	sourceObject: any,
	processedObject: any,
) => ValidatorResult;

/**
 * LLM processor for JSON objects that support object slicing
 *
 * JSON object will be split to chunks that will be processed one by one
 */
export class LLMJsonProcessor {
	constructor(
		private readonly fetcher: LLMFetcher,
		private readonly config: {
			concurrency?: number;
			termsLimit?: number;
			chunkParsingRetriesLimit?: number;
			backpressureTimeout?: {
				base: number;
				max: number;
			} | null;
		} = {},
	) {}

	public async process<T extends {}>(
		sourceObject: T,
		{
			prompt,
			onParsingError,
			filter,
			validate,
			validateSlice,
		}: {
			prompt: (json: string) => string;
			onParsingError?: ParsingErrorFixer;
			filter?: ObjectFilter;
			validate?: ObjectTransformingValidator;
			validateSlice?: ObjectTransformingValidator;
		},
	) {
		const filteredObject = filter
			? splitObjectByFilter(sourceObject, filter)
			: { included: sourceObject, excluded: {} };

		const objectSlices = sliceJsonString(
			JSON.stringify(filteredObject.included),
			this.fetcher.getLengthLimit(),
			this.config.termsLimit ?? 6,
		);

		const transformedSlices: [string, unknown][][] = Array(objectSlices.length);

		let index = 0;
		const abort = new AbortController();
		await Promise.all(
			Array(this.config.concurrency ?? 3)
				.fill(null)
				.map(async () => {
					while (true) {
						if (abort.signal.aborted) return;

						const currentIndex = index++;
						const slice = objectSlices[currentIndex];

						// End of worker
						if (!slice) return;

						const messages: MessageObject[] = [
							{ role: 'user', content: prompt(slice) },
						];
						let correctionMessages: null | MessageObject[] = null;
						for (let retry = 0; true; retry++) {
							if (abort.signal.aborted) return;

							try {
								// Update request
								if (correctionMessages) {
									messages.push(...correctionMessages);
									correctionMessages = null;
								}

								const response = await this.fetcher.query(
									structuredClone(messages),
								);

								// Update messages
								messages.push(...response);

								const transformedSlice = messages.slice(-1)[0].content;
								if (!transformedSlice)
									throw new TypeError('Empty message in response');

								const sourceObject = JSON.parse(slice);
								let transformedObject;
								try {
									transformedObject = JSON.parse(transformedSlice);
								} catch (error) {
									if (onParsingError) {
										correctionMessages =
											onParsingError(transformedSlice);
									}
									throw error;
								}

								// Validate slice
								if (validateSlice) {
									const validationResult = validateSlice(
										sourceObject,
										transformedObject,
									);

									if (!validationResult.isValid) {
										// Correction messages
										if (validationResult.correctionRequest) {
											correctionMessages =
												validationResult.correctionRequest;
										}

										throw validationResult.reason
											? validationResult.reason
											: new TypeError('Invalid slice');
									}
								}

								transformedSlices[currentIndex] =
									Object.entries(transformedObject);
								break;
							} catch (error) {
								if (
									retry++ < (this.config.chunkParsingRetriesLimit ?? 5)
								) {
									if (this.config.backpressureTimeout !== null) {
										await waitTimeWithJitter({
											base: 100,
											max: 1000,

											// Override by user config
											...this.config.backpressureTimeout,

											retry,
										});
									}

									continue;
								}

								throw error;
							}
						}
					}
				}),
		).catch((error) => {
			abort.abort(error);
			throw error;
		});

		const transformedObject = deepmerge.all([
			Object.fromEntries(transformedSlices.flat()),
			filteredObject.excluded,
		]);

		// Validate result object
		if (validate) {
			const validationResult = validate(sourceObject, transformedObject);
			if (!validationResult.isValid)
				throw validationResult.reason ?? new TypeError('Invalid result object');
		}

		return transformedObject;
	}
}
