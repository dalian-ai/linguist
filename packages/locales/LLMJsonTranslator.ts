import traverse from 'traverse';

import { MessageObject } from './LLMFetcher';
import { LLMJsonProcessor } from './LLMJsonProcessor';
import { getObjectPatch } from './utils/json';

export type TranslationContext = {
	json: Record<string, any>;
	from: string;
	to: string;
};

export type InvalidTranslationPromptFetcher = (data: {
	missedPaths: string[];
	addedPaths: string[];
	context: TranslationContext;
}) => MessageObject[];

export const createValidator =
	({
		fix,
		context,
	}: {
		context: TranslationContext;
		fix?: InvalidTranslationPromptFetcher;
	}) =>
		(source: Record<any, any>, transformed: Record<any, any>) => {
			const missedPaths = traverse(getObjectPatch(source, transformed).superset)
				.paths()
				.slice(1)
				.map((path) => path.join('.'));
			const addedPaths = traverse(getObjectPatch(transformed, source).superset)
				.paths()
				.slice(1)
				.map((path) => path.join('.'));

			if (missedPaths.length === 0 && addedPaths.length === 0) return { isValid: true };

			return {
				isValid: false,
				reason: new Error('Not equal structures'),
				correctionRequest: fix
					? fix({
						missedPaths,
						addedPaths,
						context: structuredClone(context),
				  })
					: undefined,
			};
		};

export class LLMJsonTranslator {
	constructor(
		private readonly jsonProcessor: LLMJsonProcessor,
		private readonly prompts: {
			translate: (json: string, from: string, to: string) => string;
			fix?: InvalidTranslationPromptFetcher;
		},
	) {}

	public async translate<T extends {}>(sourceObject: T, from: string, to: string) {
		const validator = createValidator({
			fix: this.prompts.fix,
			context: {
				json: structuredClone(sourceObject),
				from,
				to,
			},
		});

		return this.jsonProcessor.process(sourceObject, {
			prompt: (json) => this.prompts.translate(json, from, to),
			validate: validator,
			validateSlice: validator,
		});
	}
}
