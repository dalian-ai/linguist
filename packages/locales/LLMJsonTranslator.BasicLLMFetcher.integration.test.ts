import 'dotenv/config';

import { BasicLLMFetcher } from './BasicLLMFetcher';
import { LLMJsonProcessor } from './LLMJsonProcessor';
import { LLMJsonTranslator } from './LLMJsonTranslator';
import dataSample from './sample.json';
import { getJsonTranslationPrompt } from './utils/prompts';

expect.addSnapshotSerializer({
	test(val) {
		return (
			val &&
			Object.prototype.hasOwnProperty.call(val, 'message') &&
			val.message !== '*REDACTED*'
		);
	},
	serialize(val, config, indentation, depth, refs, printer) {
		return printer(
			{
				...val,
				message: '*REDACTED*',
			},
			config,
			indentation,
			depth,
			refs,
		);
	},
});

test('Translation for a sample localization file', async () => {
	const translator = new LLMJsonTranslator(
		new LLMJsonProcessor(
			new BasicLLMFetcher(
				{
					apiKey: process.env.OPENAI_API_KEY as string,
					baseURL: process.env.OPENAI_BASE_URL,
					dangerouslyAllowBrowser: true,
				},
				{
					model: 'openai/gpt-4.1-mini',
					temperature: 0,
				},
			),
			{ concurrency: 10 },
		),
		{
			translate: getJsonTranslationPrompt,
		},
	);

	await expect(translator.translate(dataSample, 'en', 'ru')).resolves.toMatchSnapshot();
}, 120_000);
