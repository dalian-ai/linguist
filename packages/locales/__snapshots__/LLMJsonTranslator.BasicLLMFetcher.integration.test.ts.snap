// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Translation for a sample localization file 1`] = `
{
  "common_action_addToDictionary": {
    "message": "*REDACTED*",
  },
  "common_action_removeFromDictionary": {
    "message": "*REDACTED*",
  },
  "common_bgUnavailable": {
    "message": "*REDACTED*",
  },
  "common_cancel": {
    "message": "*REDACTED*",
  },
  "common_close": {
    "message": "*REDACTED*",
  },
  "common_copy": {
    "message": "*REDACTED*",
  },
  "common_error": {
    "message": "*REDACTED*",
  },
  "common_offlineTranslator": {
    "message": "*REDACTED*",
  },
  "common_retry": {
    "message": "*REDACTED*",
  },
  "common_translate": {
    "message": "*REDACTED*",
  },
  "component_hotkey_placeholder": {
    "message": "*REDACTED*",
  },
  "component_hotkey_recordPlaceholder": {
    "message": "*REDACTED*",
  },
  "component_hotkey_resetButton": {
    "message": "*REDACTED*",
  },
  "contextMenu_disablePageTranslation": {
    "message": "*REDACTED*",
  },
  "contextMenu_translatePage": {
    "message": "*REDACTED*",
  },
  "contextMenu_translateSelectedText": {
    "message": "*REDACTED*",
  },
  "dictionaryButton_add": {
    "message": "*REDACTED*",
  },
  "dictionaryButton_delete": {
    "message": "*REDACTED*",
  },
  "dictionary_button_export": {
    "message": "*REDACTED*",
  },
  "dictionary_button_removeAll": {
    "message": "*REDACTED*",
  },
  "dictionary_deleteAll": {
    "message": "*REDACTED*",
  },
  "dictionary_deleteConfirmation": {
    "message": "*REDACTED*",
  },
  "dictionary_description": {
    "message": "*REDACTED*",
  },
  "dictionary_emptyDictionary": {
    "message": "*REDACTED*",
  },
  "dictionary_filter_from": {
    "message": "*REDACTED*",
  },
  "dictionary_filter_to": {
    "message": "*REDACTED*",
  },
  "dictionary_message_deleteAll_success": {
    "message": "*REDACTED*",
  },
  "dictionary_notFoundEntries": {
    "message": "*REDACTED*",
  },
  "dictionary_pageTitle": {
    "message": "*REDACTED*",
  },
  "dictionary_resetFilters": {
    "message": "*REDACTED*",
  },
  "dictionary_searchPlaceholder": {
    "message": "*REDACTED*",
  },
  "editorWindow_close": {
    "message": "*REDACTED*",
  },
  "editorWindow_data_code": {
    "message": "*REDACTED*",
  },
  "editorWindow_data_name": {
    "message": "*REDACTED*",
  },
  "editorWindow_message_invalidName": {
    "message": "*REDACTED*",
  },
  "editorWindow_save": {
    "message": "*REDACTED*",
  },
  "ext_description": {
    "message": "*REDACTED*",
  },
  "ext_name": {
    "message": "*REDACTED*",
  },
  "history_control_selectEntry": {
    "message": "*REDACTED*",
  },
  "history_controls_clearHistory": {
    "message": "*REDACTED*",
  },
  "history_controls_deleteSelected": {
    "message": "*REDACTED*",
  },
  "history_controls_selectAll": {
    "message": "*REDACTED*",
  },
  "history_message_clearHistoryConfirmation": {
    "message": "*REDACTED*",
  },
  "history_message_deleteEntryConfirmation": {
    "message": "*REDACTED*",
  },
  "history_message_deleteSelectedEntriesConfirmation": {
    "message": "*REDACTED*",
  },
  "history_message_entriesNotFound": {
    "message": "*REDACTED*",
  },
  "history_message_noEntries": {
    "message": "*REDACTED*",
  },
  "history_notification_disabled": {
    "message": "*REDACTED*",
  },
  "history_pageTitle": {
    "message": "*REDACTED*",
  },
  "history_searchPlaceholder": {
    "message": "*REDACTED*",
  },
  "inlineTranslator_showOriginalText": {
    "message": "*REDACTED*",
  },
  "lang_change": {
    "message": "*REDACTED*",
  },
  "lang_detect": {
    "message": "*REDACTED*",
  },
  "lang_select": {
    "message": "*REDACTED*",
  },
  "lang_swap": {
    "message": "*REDACTED*",
  },
  "message_askResetConfig": {
    "message": "*REDACTED*",
  },
  "message_cacheFlushed": {
    "message": "*REDACTED*",
  },
  "message_changesSaved": {
    "message": "*REDACTED*",
  },
  "message_error": {
    "message": "*REDACTED*",
  },
  "message_errorOfTranslate": {
    "message": "*REDACTED*",
  },
  "message_serviceRequiredKey": {
    "message": "*REDACTED*",
  },
  "message_translateOfText": {
    "message": "*REDACTED*",
  },
  "message_unknownError": {
    "message": "*REDACTED*",
  },
  "pageTranslator_alwaysTranslateLang": {
    "message": "*REDACTED*",
  },
  "pageTranslator_commonPreferences_autoTranslate_disable": {
    "message": "*REDACTED*",
  },
  "pageTranslator_commonPreferences_autoTranslate_disableForAll": {
    "message": "*REDACTED*",
  },
  "pageTranslator_commonPreferences_autoTranslate_enable": {
    "message": "*REDACTED*",
  },
  "pageTranslator_commonPreferences_title": {
    "message": "*REDACTED*",
  },
  "pageTranslator_option_autoTranslate": {
    "message": "*REDACTED*",
  },
  "pageTranslator_showOptions": {
    "message": "*REDACTED*",
  },
  "pageTranslator_showOriginal": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_always": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_alwaysForThisLang": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_default": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_defaultForThisLang": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_never": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_autoTranslate_neverForThisLang": {
    "message": "*REDACTED*",
  },
  "pageTranslator_sitePreferences_title": {
    "message": "*REDACTED*",
  },
  "pageTranslator_translatePage": {
    "message": "*REDACTED*",
  },
  "pageTranslator_translationReport": {
    "message": "*REDACTED*",
  },
  "pageTranslator_translationReport_queue": {
    "message": "*REDACTED*",
  },
  "pageTranslator_translationReport_reject": {
    "message": "*REDACTED*",
  },
  "pageTranslator_translationReport_resolve": {
    "message": "*REDACTED*",
  },
  "popup_tabIdNotFound": {
    "message": "*REDACTED*",
  },
  "popup_tab_translatePage": {
    "message": "*REDACTED*",
  },
  "popup_tab_translateText": {
    "message": "*REDACTED*",
  },
  "settings_button_cancel": {
    "message": "*REDACTED*",
  },
  "settings_button_export": {
    "message": "*REDACTED*",
  },
  "settings_button_import": {
    "message": "*REDACTED*",
  },
  "settings_button_reset": {
    "message": "*REDACTED*",
  },
  "settings_button_saveChanges": {
    "message": "*REDACTED*",
  },
  "settings_message_clearCache_success": {
    "message": "*REDACTED*",
  },
  "settings_message_common_invalidValueType": {
    "message": "*REDACTED*",
  },
  "settings_message_importConfig_invalidFile": {
    "message": "*REDACTED*",
  },
  "settings_message_importConfig_success": {
    "message": "*REDACTED*",
  },
  "settings_message_resetConfig_confirm": {
    "message": "*REDACTED*",
  },
  "settings_message_resetConfig_success": {
    "message": "*REDACTED*",
  },
  "settings_message_saveChanges_success": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon_item_auto": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon_item_color": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon_item_dark": {
    "message": "*REDACTED*",
  },
  "settings_option_appIcon_item_light": {
    "message": "*REDACTED*",
  },
  "settings_option_cache": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_clear": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_clear_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_enable": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_enable_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_ignoreCase": {
    "message": "*REDACTED*",
  },
  "settings_option_cache_ignoreCase_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_commonSettings": {
    "message": "*REDACTED*",
  },
  "settings_option_customTranslatorModule": {
    "message": "*REDACTED*",
  },
  "settings_option_customTranslatorModule_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_customTranslatorModule_manageButton": {
    "message": "*REDACTED*",
  },
  "settings_option_history_enable": {
    "message": "*REDACTED*",
  },
  "settings_option_history_enable_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_mainPreferences": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_detectLanguageByContent": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_detectLanguageByContent_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_enableContextMenu": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_enableContextMenu_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_excludeSelectors": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_excludeSelectors_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_lazyTranslate": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_lazyTranslate_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_originalTextPopup": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_originalTextPopup_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_toggleTranslationHotkey": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_toggleTranslationHotkey_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_translatableAttributes": {
    "message": "*REDACTED*",
  },
  "settings_option_pageTranslation_translatableAttributes_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_popupTab_pageTranslator_showCounters": {
    "message": "*REDACTED*",
  },
  "settings_option_popupTab_pageTranslator_showCounters_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_detectedLangFirst": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_detectedLangFirst_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_disableWhileTranslatePage": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_enable": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_focusOnTranslateButton": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_focusOnTranslateButton_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_header_languageChoice": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_isUseAutoForDetectLang": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_isUseAutoForDetectLang_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_mode": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_mode_item_contextMenu": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_mode_item_popupButton": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_mode_item_quickTranslate": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers_key_alt": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers_key_ctrl": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers_key_meta": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_modifiers_key_shift": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_rememberDirection": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_showOnceForSelection": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_showOnceForSelection_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_showOriginalText": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_showOriginalText_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_strictSelection": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_strictSelection_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_timeoutForHideButton": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_timeoutForHideButton_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_zIndex": {
    "message": "*REDACTED*",
  },
  "settings_option_selectTranslation_zIndex_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_rememberText": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_rememberText_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_spellCheck": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_suggestLanguage": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_suggestLanguageAlways": {
    "message": "*REDACTED*",
  },
  "settings_option_textTranslator_suggestLanguageAlways_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_translatePreferences": {
    "message": "*REDACTED*",
  },
  "settings_option_translateScheduler": {
    "message": "*REDACTED*",
  },
  "settings_option_translateScheduler_delay": {
    "message": "*REDACTED*",
  },
  "settings_option_translateScheduler_delay_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_translateScheduler_retryLimit": {
    "message": "*REDACTED*",
  },
  "settings_option_translateScheduler_retryLimit_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_translatorModule": {
    "message": "*REDACTED*",
  },
  "settings_option_translatorModule_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_tts": {
    "message": "*REDACTED*",
  },
  "settings_option_ttsCustomModules": {
    "message": "*REDACTED*",
  },
  "settings_option_ttsCustomModules_button": {
    "message": "*REDACTED*",
  },
  "settings_option_ttsCustomModules_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_ttsModule": {
    "message": "*REDACTED*",
  },
  "settings_option_ttsModule_desc": {
    "message": "*REDACTED*",
  },
  "settings_option_userLanguage": {
    "message": "*REDACTED*",
  },
  "settings_option_userLanguage_desc": {
    "message": "*REDACTED*",
  },
  "settings_pageTitle": {
    "message": "*REDACTED*",
  },
  "settings_section_history": {
    "message": "*REDACTED*",
  },
  "settings_section_popup": {
    "message": "*REDACTED*",
  },
  "textTranslator_suggestLanguage": {
    "message": "*REDACTED*",
  },
  "textTranslator_translateInputPlaceholder": {
    "message": "*REDACTED*",
  },
  "textTranslator_translatePlaceholder": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_add": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_close": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_emptyTranslatorsListText": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_message_translatorRemovingConfirmation": {
    "message": "*REDACTED*",
    "placeholders": {
      "translator_name": {
        "content": "$1",
        "example": "Google translator",
      },
    },
  },
  "translatorsManagerWindow_title": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_translator_delete": {
    "message": "*REDACTED*",
  },
  "translatorsManagerWindow_translator_edit": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_add": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_close": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_emptyTranslatorsListText": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_message_ttsRemoveConfirmation": {
    "message": "*REDACTED*",
    "placeholders": {
      "tts_name": {
        "content": "$1",
        "example": "Google TTS",
      },
    },
  },
  "ttsManagerWindow_title": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_translator_delete": {
    "message": "*REDACTED*",
  },
  "ttsManagerWindow_translator_edit": {
    "message": "*REDACTED*",
  },
  "validation_number": {
    "message": "*REDACTED*",
  },
  "validation_numberBetweenRange": {
    "message": "*REDACTED*",
  },
  "validation_positiveNumber": {
    "message": "*REDACTED*",
  },
}
`;
